writeTextFile,doc/build-helpers: assert destination without a leading slash

networkException opened the pull request:
## Description of changes

This pull request improves the behavior of the `writeTextFile` trivial builder by emitting
a proper error message when a destination without a leading `/` is passed. Alongside this,
a few wrongly documented calls in the docs have been fixed.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

networkException committed:
doc/build-helpers: fix wrong invokations of writeTextFile with destination

This patch fixes occurances of writeTextFile invokations with
a destination set but without a leading slash. This would cause
an opaque build time error.

networkException committed:
writeTextFile: assert destination starting with a /

This patch adds an assertion to the writeTextFile trivial
builder, asserting that a specified destination starts with
a slash.

This matches the current behavior of the function and merely
provides a helpful error message.

emilylange commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -86,6 +86,10 @@ rec {
     , preferLocalBuild ? true
     , derivationArgs ? { }
     }:
+    assert lib.assertMsg (destination == "" || lib.hasPrefix "/" destination && destination != "/") ''
```
Could also be written with our logical implication operator, which appears to be slightly more common for such asserts.

Not blocking.

```suggestion
    assert lib.assertMsg (destination != "" -> (lib.hasPrefix "/" destination && destination != "/")) ''
```

Or, if we want to be really fancy:

```suggestion
    assert lib.assertMsg (destination != "" -> (lib.hasPrefix "/" destination)) ''
      destination must start with / if provided
    '';
    assert lib.assertMsg (destination != "/") ''
      destination cannot be /
    '';
```

networkException replied:
I've kept this as one assert for now and updated the wording to make the `/` case a little more clear

emilylange commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -86,6 +86,10 @@ rec {
     , derivationArgs ? { }
     }:
+    assert lib.assertMsg (destination == "" || lib.hasPrefix "/" destination && destination != "/") ''
+      destination must start with a /
```
Nit, not blocking

```suggestion
      destination must start with / if provided
```

emilylange reviewed (commented):
We don't have a lot of tests for this (just 2) but for what it's worth `tests.trivial-builders.writeTextFile` seems to pass as is.

Feels like a no-brainer to me, lgtm.

Two suggestions from me, both non-blocking.

drupol commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -86,6 +86,10 @@ rec {
     , derivationArgs ? { }
     }:
+    assert lib.assertMsg (destination == "" || lib.hasPrefix "/" destination && destination != "/") ''
+      destination must start with a /
```
Would it make sense to name it an "absolute path" then ?

From [here](https://www.redhat.com/sysadmin/linux-path-absolute-relative):

> An absolute path always begins from the absolute start of your hard drive and describes every step you must take through the filesystem to end up at the target location.

Here's my suggestion:

```suggestion
      destination must be an absolute path, starting from the root directory (typically with a forward slash /). Ensure that the full directory structure is specified, and avoid using relative paths.
```

networkException replied:
I'm not sure if that wording might not be misleading here. Destination is always relative to its store path (by just setting `target=$out${lib.escapeShellArg destination}`).

Also I'd opt for stronger wording than "avoid". This isn't a lint for best practice but rather a way to catch malformed input at eval time rather than having the derivation fail to build.

I'm open for making the message more descriptive though, perhaps something like this (still not quite happy about bending the definition of "absolute" though)

```suggestion
      destination must be an absolute path, relative to the derivation's out path. Ensure that the path starts with a / and specifies at least the filename.
```

drupol replied:
Agreed ! Perfect!

networkException commented:
I've updated the error message as per https://github.com/NixOS/nixpkgs/pull/343595#discussion_r1770306919 which also now includes the offending input printed back.

I've also updated the check expression to use the logical implication operator as per https://github.com/NixOS/nixpkgs/pull/343595#discussion_r1769678070

emilylange commented:
@ofborg eval
@ofborg build tests.trivial-builders
